pkg/utils/tsoutil/tso_dispatcher.go

```diff
@@ -45,119 +43,159 @@ type TSODispatcher struct {
 	tsoProxyHandleDuration prometheus.Histogram
 	tsoProxyBatchSize      prometheus.Histogram
 
+	ctx context.Context
 	// dispatchChs is used to dispatch different TSO requests to the corresponding forwarding TSO channels.
-	dispatchChs sync.Map // Store as map[string]chan Request
+	dispatchChs sync.Map // Store as map[string]chan Request (forwardedHost -> dispatch channel)
+	// lastErrors is used to record the last error of each forwarding TSO channel.
+	lastErrors sync.Map // Store as map[string]error (forwardedHost -> last error)
 }
 
 // NewTSODispatcher creates and returns a TSODispatcher
-func NewTSODispatcher(tsoProxyHandleDuration, tsoProxyBatchSize prometheus.Histogram) *TSODispatcher {
+func NewTSODispatcher(
+	ctx context.Context, tsoProxyHandleDuration, tsoProxyBatchSize prometheus.Histogram,
+) *TSODispatcher {
 	tsoDispatcher := &TSODispatcher{
+		ctx:                    ctx,
 		tsoProxyHandleDuration: tsoProxyHandleDuration,
 		tsoProxyBatchSize:      tsoProxyBatchSize,
 	}
 	return tsoDispatcher
 }
 
-// DispatchRequest is the entry point for dispatching/forwarding a tso request to the detination host
-func (s *TSODispatcher) DispatchRequest(
-	ctx context.Context,
-	req Request,
-	tsoProtoFactory ProtoFactory,
-	doneCh <-chan struct{},
-	errCh chan<- error,
-	tsoPrimaryWatchers ...*etcdutil.LoopWatcher) {
+// GetAndDeleteLastError gets and deletes the last error of the forwarded host
+func (s *TSODispatcher) GetAndDeleteLastError(forwardedHost string) error {
+	if val, loaded := s.lastErrors.LoadAndDelete(forwardedHost); loaded {
+		return val.(error)
+	}
+	return nil
+}
+
+// DispatchRequest is the entry point for dispatching/forwarding a tso request to the destination host
+func (s *TSODispatcher) DispatchRequest(req Request, tsoProtoFactory ProtoFactory) {
 	val, loaded := s.dispatchChs.LoadOrStore(req.getForwardedHost(), make(chan Request, maxMergeRequests))
 	reqCh := val.(chan Request)
 	if !loaded {
-		tsDeadlineCh := make(chan deadline, 1)
-		go s.dispatch(ctx, tsoProtoFactory, req.getForwardedHost(), req.getClientConn(), reqCh, tsDeadlineCh, doneCh, errCh, tsoPrimaryWatchers...)
-		go watchTSDeadline(ctx, tsDeadlineCh)
+		go s.startDispatchLoop(req.getForwardedHost(), req.getClientConn(), reqCh, tsoProtoFactory)
 	}
 	reqCh <- req
 }
 
-func (s *TSODispatcher) dispatch(
-	ctx context.Context,
-	tsoProtoFactory ProtoFactory,
-	forwardedHost string,
-	clientConn *grpc.ClientConn,
-	tsoRequestCh <-chan Request,
-	tsDeadlineCh chan<- deadline,
-	doneCh <-chan struct{},
-	errCh chan<- error,
-	tsoPrimaryWatchers ...*etcdutil.LoopWatcher) {
+// startDispatchLoop starts the dispatch loop for the forwarded host
+func (s *TSODispatcher) cleanup(
+	forwardedHost string, finalForwardErr error,
+	unprocessedRequests []Request, unprocessedReqCount int,
+) {
+	pendingReqCount := unprocessedReqCount
+	pendingRequests := unprocessedRequests[:unprocessedReqCount]
```

Comment:
Why not just use the parameter passed in the function?

Reply:
you are right